Optional argument to convert multiple TextRuns into an expression

markgould opened the pull request:
This PR adds an optional argument /ssrs:MultipleTextRunBehavior=CombinedExpression (Defaults to RichText) to convert multiple TextRuns into a combined expression

In our reports, we have several data fields that look like this:

```
<Textbox Name="szAddress1">
  <CanGrow>true</CanGrow>
  <KeepTogether>true</KeepTogether>
  <Paragraphs>
    <Paragraph>
      <TextRuns>
        <TextRun>
          <Value>=Fields!Address1.Value</Value>
          <Style>
            <FontSize>7pt</FontSize>
          </Style>
        </TextRun>
        <TextRun>
          <Value xml:space="preserve"> </Value>
          <Style>
            <FontSize>7pt</FontSize>
          </Style>
        </TextRun>
        <TextRun>
          <Value>=Fields!Address2.Value</Value>
          <Style>
            <FontSize>7pt</FontSize>
          </Style>
        </TextRun>
        <TextRun>
          <Value xml:space="preserve"> </Value>
          <Style>
            <FontSize>7pt</FontSize>
          </Style>
        </TextRun>
        <TextRun>
          <Value>=Fields!City.Value</Value>
          <Style>
            <FontSize>7pt</FontSize>
          </Style>
        </TextRun>
        <TextRun>
          <Value xml:space="preserve"> </Value>
          <Style>
            <FontSize>7pt</FontSize>
          </Style>
        </TextRun>
        <TextRun>
          <Value>=Fields!State.Value</Value>
          <Style>
            <FontSize>7pt</FontSize>
          </Style>
        </TextRun>
        <TextRun>
          <Value xml:space="preserve"> </Value>
          <Style>
            <FontSize>7pt</FontSize>
          </Style>
        </TextRun>
        <TextRun>
          <Value>=Fields!PostalCode.Value</Value>
          <Style>
            <FontSize>7pt</FontSize>
          </Style>
        </TextRun>
      </TextRuns>
      <Style />
    </Paragraph>
  </Paragraphs>
  <rd:DefaultName>szAddress1</rd:DefaultName>
  <Style>
    <Border>
      <Color>LightGrey</Color>
      <Style>Solid</Style>
    </Border>
    <VerticalAlign>Middle</VerticalAlign>
    <PaddingLeft>2pt</PaddingLeft>
    <PaddingRight>2pt</PaddingRight>
    <PaddingTop>2pt</PaddingTop>
    <PaddingBottom>2pt</PaddingBottom>
  </Style>
</Textbox>
```
Currently this gets converted to a a rich text control with content of:

> Cannot convert the 'szAddress1' RichText with the following Run Expression: '[Address1]'. Cannot convert the 'szAddress1'  RichText with the following Run Expression: '[Address2]'. Cannot convert the 'szAddress1' RichText with the following Run  Expression: '[City]'. Cannot convert the 'szAddress1' RichText with the following Run Expression: '[State]'. Cannot convert the 'szAddress1' RichText with the following Run Expression: '[PostalCode]'.

With this PR, using the new argument, it now creates an expression of:
> [Address1] + ' ' + [Address2] + ' ' + [City] + ' ' + [State] + ' ' + [PostalCode]

I know you mentioned using the mail merge, which I can look at if you'd like, but this does create expressions like other data bound fields.

**Before:**
![image](https://user-images.githubusercontent.com/2712328/104319552-a4bbaf80-54a6-11eb-8d4a-e7a513111e65.png)

**After:**
![image](https://user-images.githubusercontent.com/2712328/104319596-b56c2580-54a6-11eb-9b54-24eea12d380b.png)

**Designer view:**
![image](https://user-images.githubusercontent.com/2712328/104318816-8bfeca00-54a5-11eb-9668-70bb7249020c.png)

markgould committed:
Add new option to combine text runs into a single expression

markgould committed:
Fix up variable name after rebase

k0st1x commented:
you are generating useful code.

but i fear that functional with new CombinedExpression key will ignore the rich formatting
https://github.com/DevExpress/Reporting.Import/pull/13/files#diff-d5b8174d6aa94b6d3575d32a9ac61e12814bd4acd676601d8a53e4d8f451e6e7R351
```cs
            if(runs.Count() > 1 && MultipleTextRunBehavior == MultipleTextRunBehavior.RichText) {
                ProcessTextBoxAsRichText(textBoxElement, container, yBodyOffset);
	                return;
            }
```
so no rich with new key.

the better option can be some logic that detects are the runs have same styles.
so it is possible to collect not only "values", but "XElement etalonStyle" and compare current run style with "etalonStyle" (which is the first run style), so if all styles are the same, execute your new logic without a switch key.

can you update the code? or would it take too much effort?

also while writing code please follow the project style rules - no spaces between operator & round bracket, and no new line for opening curly bracket.

markgould commented:
I can look at making those changes. I can fix the code style, I think ReSharper has done that automatically

markgould committed:
Remove Rich Text behavior parameter and auto detect styles. Follow code formatting styles.

markgould committed:
Correct formatting

markgould committed:
Remove additional whitespace

markgould commented:
I was able to have ReSharper auto detect the styles, and reformatted accordingly. 

I added logic to check how many distinct styles are in the TextRuns. If more than one, it will use the rich text logic, otherwise it will use the new logic.

Let me know what you think!

k0st1x commented on Import/ReportingServicesConverter.cs:
```diff
@@ -345,10 +346,13 @@ public class ReportingServicesConverter : ExternalConverterBase, IReportingServi
-                return;
+            List<XElement> runs = textBoxElement.Descendants(xmlns + "TextRun").ToList();
+            if(runs.Count > 1) {
+                var distinctStyles = runs.Descendants(xmlns + "Style").Select(e => e.ToString()).Distinct().Count();
```
please use the
```cs
var distinctStylesCount = runs.Descendants(xmlns + "Style").Distinct(XNode.EqualityComparer).Count();
```
instead

markgould replied:
Oh, nice. Wasn't familiar with that equality comparer. That cleans it up, thanks!

k0st1x commented on Import/ReportingServicesConverter.cs:
```diff
@@ -546,35 +546,47 @@ public class ReportingServicesConverter : ExternalConverterBase, IReportingServi
+                        sb.Append("=");
+                        for (var i = 0; i < values.Count; i++) {
+                            var value = values[i];
+                            sb.Append(IsExpression(value) ? value.Substring(1) : $"\"{value}\"");
```
value should be escaped: double quotes to quadruple quotes :) " -> ""

markgould replied:
That would not compile. The correct way to escape quotes in interpolated strings is the \ character.

```
var value = "test";
Console.Write($"\"{value}\"");
```

Writes "test" to the console

k0st1x replied:
```cs
$"\"{value.Replace("\"", "\"\"")}\""
```

markgould replied:
Aha! Sorry, I misunderstood your request. Fixed!

markgould committed:
Use XNode EqualityComparer

markgould committed:
Escape quotes in values
